src/main/scala/is/hail/utils/UInt.scala

```diff
@@ -46,25 +50,57 @@ class UInt private(val i: Int) extends AnyVal {
     if (l == r) 0 else if (l > r) 1 else -1
   }
 
-  def +(right: UInt): UInt = UInt(toLong + right.toLong)
-
-  def -(right: UInt): UInt = UInt(toLong - right.toLong)
-
-  def *(right: UInt): UInt = UInt(toLong * right.toLong)
-
-  def /(right: UInt): UInt = UInt(toLong / right.toLong)
-
-  def <=(right: UInt): Boolean = toLong <= right.toLong
-
-  def >=(right: UInt): Boolean = toLong >= right.toLong
-
-  def <(right: UInt): Boolean = toLong < right.toLong
-
-  def >(right: UInt): Boolean = toLong > right.toLong
-
-  def ==(right: UInt): Boolean = toLong == right.toLong
-
-  def !=(right: UInt): Boolean = toLong != right.toLong
+  def intRep: Int = i
+
+  def +(right: UInt)(implicit uct: ClassTag[UInt]): UInt = UInt(toLong + right.toLong)
+  def +(right: Int): UInt = UInt(toLong + right)
+  def +(right: Double): Double = toDouble + right
+  def +(right: Long): Long = toLong + right
+
+  def -(right: UInt)(implicit uct: ClassTag[UInt]): UInt = UInt(toLong - right.toLong)
+  def -(right: Int): UInt = UInt(toLong - right)
+  def -(right: Double): Double = toDouble - right
+  def -(right: Long): Long = toLong - right
+
+  def *(right: UInt)(implicit uct: ClassTag[UInt]): UInt = UInt(toLong * right.toLong)
+  def *(right: Int): UInt = UInt(toLong * right)
+  def *(right: Double): Double = toDouble * right
+  def *(right: Long): Long = toLong * right
+
+  def /(right: UInt)(implicit uct: ClassTag[UInt]): UInt = UInt(toLong / right.toLong)
+  def /(right: Int): UInt = UInt(toLong / right)
+  def /(right: Double): Double = toDouble / right
+  def /(right: Long): Long = toLong / right
+
+  def <=(right: UInt)(implicit uct: ClassTag[UInt]): Boolean = toLong <= right.toLong
+  def <=(right: Int): Boolean = toLong <= right
+  def <=(right: Double): Boolean = toDouble <= right
+  def <=(right: Long): Boolean = toLong <= right
+
+  def >=(right: UInt)(implicit uct: ClassTag[UInt]): Boolean = toLong >= right.toLong
+  def >=(right: Int): Boolean = toLong >= right
+  def >=(right: Double): Boolean = toDouble >= right
+  def >=(right: Long): Boolean = toLong >= right
+
+  def <(right: UInt)(implicit uct: ClassTag[UInt]): Boolean = toLong < right.toLong
+  def <(right: Int): Boolean = toLong < right
+  def <(right: Double): Boolean = toDouble < right
+  def <(right: Long): Boolean = toLong < right
+
+  def >(right: UInt)(implicit uct: ClassTag[UInt]): Boolean = toLong > right.toLong
+  def >(right: Int): Boolean = toLong > right
+  def >(right: Double): Boolean = toDouble > right
+  def >(right: Long): Boolean = toLong > right
+
+  def ==(right: UInt)(implicit uct: ClassTag[UInt]): Boolean = toLong == right.toLong
+  def ==(right: Int): Boolean = toLong == right.toLong
+  def ==(right: Double): Boolean = toDouble == right
+  def ==(right: Long): Boolean = toLong == right
+
+  def !=(right: UInt)(implicit uct: ClassTag[UInt]): Boolean = toLong != right.toLong
```

Comment:
OK, the error is: double definition: functions have same type after erasure

So any change that changes the Int vs UInt functions type should work.  Here is a test:

```
class UInt(val i: Int) extends AnyVal

object X {
  def f(i: Int): Int = { println("f1"); i }
  def f(i: UInt, dummy: Boolean = false): Int = { println("f2"); i.i }

  def main(args: Array[String]) {
    f(5)
    f(new UInt(5))
  }
}
```

This prints f1 then f2.  I prefer the boolean change like above for two reasons: first, having two sets of function parameters causes a function object to be allocated, and second, since we can make any perturbation, I think we should choose the simplest, and Boolean is simpler in some sense than ClassTag.